WordPress/Classes/Utility/BackgroundTasks/WeeklyRoundupBackgroundTask.swift

```diff
@@ -135,20 +155,24 @@ private class WeeklyRoundupDataProvider {
     /// Retrieves the sites considered by Weekly Roundup for reporting.
     ///
     /// - Returns: the requested sites (could be an empty array if there's none) or an error if there is one.
-    ///
-    private func getSites(result: @escaping (Result<[Blog], Error>) -> Void) {
-
-        switch getAllSites() {
-        case .success(let sites):
-            filterCandidateSites(sites, result: result)
-        case .failure(let error):
-            result(.failure(error))
+    private func getSites(result: @escaping (Result<[Site], Error>) -> Void) {
+        self.coreDataStack.performAndSave { [weak self] context in
+            guard let self else {
+                result(.success([]))
+                return
+            }
+            switch getAllSites(in: context) {
+            case .success(let sites):
+                filterCandidateSites(sites, result: result)
+            case .failure(let error):
+                result(.failure(error))
+            }
```

Comment:
What do you think we call the `filterCandidateSites(sites, result: result)` outside of the `performAndSave` function, considering it runs asynchronously and doesn't need a context object? We can simply move the switch case out:

```swift
let allSites = coreDataStack.performQuery { self.getAllSites(in: $0) }
switch allSite {
// ...
}
```

BTW, I think it's okay to use performQuery in this case? The query traverses `Blog` models, which should finish pretty quickly considering there shouldn't be many `Blog` in the database?

Reply:
@crazytonyli Yes that should work, but I thought since this is a "background task", it would be great to execute all operations on the background. 

Also, I don't know how many blogs users have but I know, A11ns have lots of blogs ( including Matt ). Of course, even if a user has a 100 blogs, I don't that would impact the performance significantly. But I think it wouldn't hurt to fetch those blogs from the background.

I'm not sure if you've come across this post ( [Core Data Performance: 6 tips you should know](https://www.avanderlee.com/swift/core-data-performance/) ) before, but the author recommends to use background contexts as much as possible. 

That being said, is there a particular reason on why fetching those blogs on the main thread is better than fetching them from the background thread?

Reply:
> Yes that should work, but I thought since this is a "background task", it would be great to execute all operations on the background.

If I understand correctly, ["background task"](https://developer.apple.com/documentation/backgroundtasks) in here means a task that runs during the app is the background, not a task that executing in non-main threads?

But I agree. We don't necessary need to use the main context to query. That makes me think we should have an option to run `performQuery` in a background context, or outright changing it to use a background context. Anyway, that's probably a bigger and unrelated change.

> That being said, is there a particular reason on why fetching those blogs on the main thread is better than fetching them from the background thread?

Mainly because what we do in this class here are querying, and I think `performQuery` has the right semantic for it.

Reply:
I agree`performQuery` has the right semantic, that's what I asked you in this [comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/21076#discussion_r1265634985) if there is a read-only equivalent of `performAndSave`. That's because I was looking for an API that queries data on the **background thread** and the closest thing is `performAndSave`. 

> That makes me think we should have an option to run performQuery in a background context, or outright changing it to use a background context. 

That would be great. I think we should even make it standard querying data from the background thread whenever possible. It is possible that fetching on the main thread contributes to those "App Hangs" errors.

Reply:
Sounds good.

For your consideration, here is an alternative way to implement this function, where the Core Data query and using the query result are separated:

```swift
// pseudo and poorly-formatted code :-)
coreDataStack.performAndSave({ self.getAllSites(in: $0)}, completion: { allSites in
  switch allSites {
    case .success(): filterCandidateSites(...)
    case .failure(error): result(.failure(error))
  }
}, queue: .global())
```